src/test/java/com/r307/arbitrader/service/cache/ExchangeFeeCacheTest.java

```diff
@@ -37,11 +40,14 @@ public void setUp() {
 
     @Test
     public void testSetAndGet() {
-        exchangeFeeCache.setCachedFee(exchange, currencyPair, new BigDecimal("0.0025"));
-        exchangeFeeCache.setCachedFee(exchange, CurrencyPair.BTC_USD, new BigDecimal("0.0010"));
-        exchangeFeeCache.setCachedFee(exchange, CurrencyPair.ETH_USD, new BigDecimal("0.0030"));
-
-        assertEquals(Optional.of(new BigDecimal("0.0025")), exchangeFeeCache.getCachedFee(exchange, currencyPair));
+        exchangeFeeCache.setCachedFee(exchange, currencyPair, new ExchangeFee(null, new BigDecimal("0.0025")));
+        exchangeFeeCache.setCachedFee(exchange, CurrencyPair.BTC_USD, new ExchangeFee(new BigDecimal("0.0002"), new BigDecimal("0.0010")));
+        exchangeFeeCache.setCachedFee(exchange, CurrencyPair.ETH_USD, new ExchangeFee(new BigDecimal("0.0001"), new BigDecimal("0.0030")));
+        exchangeFeeCache.setCachedFee(exchange, CurrencyPair.BCC_USD, new ExchangeFee(null, new BigDecimal("0.0030")));
+
+        assertTrue(exchangeFeeCache.getCachedFee(exchange, currencyPair).isPresent());
+        assertEquals(new BigDecimal("0.0025"), exchangeFeeCache.getCachedFee(exchange, currencyPair).get().getLongFee());
+        assertFalse(exchangeFeeCache.getCachedFee(exchange, currencyPair).get().getShortFee().isPresent());
```

Comment:
I don't see a test for the short fees, and I have a concern there. I see in the example that you set the "fee" to 0.0026 and the "marginFee" to 0.0002. In `getShortFee()` you retrieve the `marginFee` but I can't seem to find where you add `fee` to it. The way I understand it, `fee` is applied to **every** trade while `marginFee` is **added** to `fee` for margin trades, for a total of 0.0028.

Reply:
Here I am only testing if the `ExchangeFeeCache` works as expected. I test only adding and retrieving to/from the cache and assert that I retrieve what I was expecting to.

I do not test trading logic here, I don't think testing trading logic belong here because `ExchangeFeeCache` has no logic regarding calculating fees but only acting as a fee caching mechanism. It only has two methods:

- setCachedFee
- computeCacheKey

Reply:
Right but I don't see any such test anywhere in your PR and I couldn't find the code where the fees are added together. It's totally possible that I just didn't see it and if that's the case please point it out.

I've read through a few times now and I'm still not seeing how we don't get a fee of 0.0002 for short trades because `getShortFee()` returns `getMarginFee()` and I don't see any place where we add `getLongFee()` to it. 

That's why I was complaining earlier about "long" and "short" fees. It's not accurate. There's a "trade" fee on every trade and an additional "margin" fee on margin trades. _Any_ trade should have fees equal to the "trade" fee. _Any_ margin trade, long or short, should have fees equal to "trade fee + margin fee". It's not related to long or short, it's related to whether you're trading on margin or not.